File: v1/src/main/java/com/api/v1/services/customer/CreditCardRegistrationServiceImpl.java
package com.api.v1.services.customer;

import com.api.v1.builders.creditcard.CreditCardBuilder;
import com.api.v1.domain.creditcard.CreditCard;
import com.api.v1.domain.creditcard.CreditCardRepository;
import com.api.v1.dtos.creditcard.CreditCardRegistrationRequestDto;
import com.api.v1.dtos.creditcard.CreditCardResponseDto;
import com.api.v1.exceptions.creditcard.DuplicatedCardNumberException;
import com.api.v1.mappers.creditcard.CreditCardResponseMapper;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
class CreditCardRegistrationServiceImpl implements CreditCardRegistrationService {

    @Autowired
    private CreditCardRepository repository;

    @Override
    public Mono<CreditCardResponseDto> register(@Valid CreditCardRegistrationRequestDto requestDto) {
        return repository
                .findAll()
                .filter(e -> e.getCardNumber().equals(requestDto.cardNumber()))
                .hasElements()
                .flatMap(exists -> {
                    if (exists) return Mono.error(new DuplicatedCardNumberException(requestDto.cardNumber()));
                    return Mono.defer(() -> {
                        CreditCard creditCard = CreditCardBuilder.create().fromDto(requestDto).build();
                        Mono<CreditCard> savedCreditCard = repository.save(creditCard);
                       return savedCreditCard.flatMap(e -> Mono.just(new CreditCardResponseMapper(e).mapToDto()));
                    });
                });
    }

}

File: v2/src/main/kotlin/com/api/v2/creditcard/services/CreditCardRegistrationServiceImpl.kt
package com.api.v2.creditcard.services

import com.api.v2.creditcard.domain.CreditCard
import com.api.v2.creditcard.domain.CreditCardRepository
import com.api.v2.creditcard.dtos.CreditCardResponseDto
import com.api.v2.creditcard.exceptions.DuplicatedCardNumberException
import com.api.v2.creditcard.mappers.CreditCardResponseMapper
import jakarta.validation.Valid
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
private class CreditCardRegistrationServiceImpl: CreditCardRegistrationService {

    @Autowired
    lateinit var creditCardRepository: CreditCardRepository

    override suspend fun register(creditCard: @Valid CreditCard): CreditCardResponseDto {
        return withContext(Dispatchers.IO) {
            val isThereADuplicatedCardNumber = creditCardRepository
                .findAll()
                .filter { e -> e.cardNumber == creditCard.cardNumber }
                .firstOrNull() != null
            if (isThereADuplicatedCardNumber) {
                throw DuplicatedCardNumberException(creditCard.cardNumber)
            }
            val savedCredit = creditCardRepository.save(creditCard)
            CreditCardResponseMapper.map(savedCredit)
        }
    }

}